Feature: Event view

infeo opened the pull request:
This PR adds an event view to the application.

The event view informs the user about events inside an unlocked vault. Such events are:
* conflict resolution (failed and successful)
* broken filesystem structure (e.g. empty `dir.c9r`)
* failed decryption (due to an invalid/tampered file)

The event view can be accessed from the main window by clicking on the bell icon in the left lower corner. The icon has an "update indicator", if new events are available. The update indicator disappears when the event window gets focus.

<img width="878" alt="pr_ready_light" src="https://github.com/user-attachments/assets/88c3af4a-29eb-4399-aecf-4325dc5d2d31" />
<img width="853" alt="pr_read_dark" src="https://github.com/user-attachments/assets/b7a2d73b-2c43-42f9-8278-fda8bb58474d" />

--------
The events are displayed in a list. Each event has a short description and a subtext mentioning the affected file (mostly encrypted version, see https://github.com/cryptomator/cryptofs/pull/278 why a decrypted path might not be available). Additionally, a counter of how often the event happend and the timestamp of last appearance is displayed. Each event offers some actions, which can be accessed by hovering over the list element and click the right button. The actions are currently "dismiss" and various forms of "reveal file" (encrypted, decrypted, original, etc), depending on the actual event.
<img width="342" alt="pr_ready_eventActions" src="https://github.com/user-attachments/assets/e4ebc111-52f8-4aaf-8a02-d55b19eb3b25" />

If a vault is locked, the event is not removed, but obfuscated. Only the affected vault is displayed.
<img width="227" alt="pr_ready_eventView_lockedEvent" src="https://github.com/user-attachments/assets/28317b85-acc5-4c3b-8aef-70dc2aaebae8" />

------
Technical details:
* events are stored inside a map keyed with (vault, some identifying path, class of event) to check if an event just reoccurred
* to not block the GUI by a damaged vault spamming constantly events, the gui list is updated at fixed intervals (every second the app polls for updates)

infeo commented on src/main/java/org/cryptomator/ui/eventview/EventListCellController.java:
```diff
@@ -0,0 +1,323 @@
+			addAction("event.brokenFileNode.copyEncrypted", () -> copyToClipboard(convertVaultPathToSystemPath(bfe.ciphertextPath()).toString()));
+		}
+		addAction("eventView.entry.brokenFileNode.copyDecrypted", () -> copyToClipboard(convertVaultPathToSystemPath(bfe.cleartextPath()).toString()));
+	}
```
Fixed in https://github.com/cryptomator/cryptomator/pull/3780/commits/6d55331ef269bdc210c6614eff383234c06a7a87

overheadhunter commented on src/main/java/org/cryptomator/event/VaultEvent.java:
```diff
@@ -0,0 +1,27 @@
+		}
+	}
+
+	public VaultEvent incrementCount(Instant timestamp) {
+		return new VaultEvent(timestamp, v, actualEvent, count+1);
+	}
```
The name suggests mutability, I would prefer a name that clearly states that `this` plus `other` will be computed into something new:

```suggestion
	public VaultEvent aggregate(VaultEvent other) {
		if (this.actualEvent.getClass().equals(other.actualEvent.getClass())) {
			throw new IllegalArgumentException("Cannot aggregate events of different types");
		}
		Instant latestTimestamp = Comparators.max(this.timestamp, other.timestamp);
		return new VaultEvent(latestTimestamp, v, actualEvent, count + 1);
	}
```

overheadhunter commented on src/main/java/org/cryptomator/event/VaultEvent.java:
```diff
@@ -0,0 +1,27 @@
+		if(timeResult != 0) {
+			return timeResult;
+		} else {
+			return this.equals(other) ? 0 : this.actualEvent.getClass().getName().compareTo(other.actualEvent.getClass().getName());
```
why the class name as second order criteria? (I suspect this is a hacky way to achieve sth in the UI)

overheadhunter commented on src/main/java/org/cryptomator/event/VaultEvent.java:
Is this supposed to only ever wrap `FileSystemEvent`s?

infeo replied:
This entry is used in the UI to gather all necessary information in one record. I refactored it, so check again.

overheadhunter commented on src/main/java/org/cryptomator/event/VaultEvent.java:
```diff
@@ -0,0 +1,27 @@
+
+import java.time.Instant;
+
+public record VaultEvent(Instant timestamp, Vault v, FilesystemEvent actualEvent, int count) implements Comparable<VaultEvent> {
```
Since this record should contain the "latest" event as a representative for all `count` events that it groups, how about
* renaming `actualEvent` to `latestEvent` or `mostRecentEvent`
* `v` to `vault`
* Moving timestamp into `FileSystemEvent`
* renaming `VaultEvent` to `FileSystemEventGroup` or `...Batch` or something

infeo replied:
Renamed in https://github.com/cryptomator/cryptomator/pull/3780/commits/893a4bcae99a0ee9a80b013e9f936a3787e26c76

overheadhunter commented on src/main/java/org/cryptomator/common/EventMap.java:
```diff
@@ -0,0 +1,160 @@
+ * The map is size restricted to {@value MAX_SIZE} elements. If a _new_ element (i.e. not already present) is added, the least recently added is removed.
+ */
+@Singleton
+public class EventMap implements ObservableMap<EventMap.EventKey, VaultEvent> {
```
* move to `event` package
* rename to `FileSystemEventRegistry` or something alike

infeo replied:
Done in https://github.com/cryptomator/cryptomator/pull/3780/commits/893a4bcae99a0ee9a80b013e9f936a3787e26c76

overheadhunter reviewed (changes requested):
The concept looks ok, but I still got a couple of question marks which indicate that the code is not self-explanatory yet (I guess mostly due to naming).

In some places there are minor inconsistencies such as passing a plaintext to a ciphertext parameter. I assume that you started with very specific events and then broadened the implementation to match further event types.

What bugs me the most is the double wrapping of FileSystemEvent into VaultEvents but then grouping based on the wrapped event's type again... This doesn't feel solid.

The EventMap already contains half the solution: The `EventKey` record. We want to group by a compound key (type and type-specific path). Symmetry dictates that the Map's value is some kind of `EventValue` (a compound event + count thingy). Looking at these, we should be able to work out a path to name stuff properly and organize data. E.g. it sparks the question why `vault` (or `v` as you call it 😉) isn't part of the compound key.

infeo commented on src/main/java/org/cryptomator/event/Answer.java:
```diff
@@ -0,0 +1,14 @@
+package org.cryptomator.event;
+
+public sealed interface Answer permits Answer.DoNothing, Answer.DoSomething {
```
Since this class is for future purposes and has no functionality, i removed it in https://github.com/cryptomator/cryptomator/pull/3780/commits/e61fb74367d3a06d9d42c1bb3c8e01bba561b0f6

infeo commented on src/main/java/org/cryptomator/event/NotificationHandler.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.ServiceLoader;
+import java.util.stream.Stream;
+
+public interface NotificationHandler {
```
Since this class is for future purposes and has no functionality, i removed it in https://github.com/cryptomator/cryptomator/commit/e61fb74367d3a06d9d42c1bb3c8e01bba561b0f6

infeo commented on src/main/java/org/cryptomator/common/vaults/Vault.java:
```diff
@@ -251,6 +259,14 @@ private void removeFromQuickAccessInternal() {
+		var wrapper = new VaultEvent(this, e);
+		Platform.runLater(() -> {
+			eventMap.put(wrapper);
+		});
```
Changed in https://github.com/cryptomator/cryptomator/pull/3780/commits/f27f3c46c1f2cfb75d85f75cb946aa047926872b

infeo commented on src/main/java/org/cryptomator/common/EventMap.java:
```diff
@@ -0,0 +1,160 @@
+@Singleton
+public class EventMap implements ObservableMap<EventMap.EventKey, VaultEvent> {
+
+	private static final int MAX_SIZE = 300;
```
As written in the PR description:
> the is size restricted to 300 entires to prevent memory issues. The oldest event is evicted if a new one enters (determined by a lru cache)

If the app runs for days and a corrupted vault is accessed regularly, we migth get memory issues.

infeo replied:
But of course, we can also simply let the map size increase, assuming that before we run into memory issues, the application is restarted.

infeo commented on src/main/java/org/cryptomator/common/EventMap.java:
```diff
@@ -0,0 +1,160 @@
+	@Override
+	public void removeListener(InvalidationListener invalidationListener) {
+		delegate.removeListener(invalidationListener);
+	}
```
Obsolete with https://github.com/cryptomator/cryptomator/pull/3780/commits/f27f3c46c1f2cfb75d85f75cb946aa047926872b

infeo commented on src/main/java/org/cryptomator/event/VaultEventsMap.java:
```diff
@@ -0,0 +1,239 @@
+			if (queueHasElements.get()) {
+				flush();
+			}
+		}, 1000, 1000, TimeUnit.MILLISECONDS);
```
What would be a good backoff strategy? Can you draft the changes?

infeo commented:
For documentation: The eventMap size restriction (and the LRU cache) will be dropped. I performed some memory test by spamming DecryptionFailedEvents (each for a different path) and even with 60_000 events the UI could be used without problem and the app heap size only reached 450MB. (see this [VisualVM Snapshot](https://github.com/user-attachments/files/19390744/eventSpam_45000Events.zip) with around 45_000 events)


Heap growth over complete app lifecycle:
![grafik](https://github.com/user-attachments/assets/d58a2081-af73-4c43-9af5-719b21e1ced1)

infeo commented on src/main/java/org/cryptomator/ui/fxapp/FxFSEventList.java:
```diff
@@ -0,0 +1,65 @@
+	private final ScheduledFuture<?> scheduledTask;
+	private final BooleanProperty unreadEvents;
+
+	@Inject
+	public FxFSEventList(FileSystemEventAggregator fsEventAggregator, ScheduledExecutorService scheduler) {
+		this.events = FXCollections.observableArrayList();
+		this.eventAggregator = fsEventAggregator;
+		this.unreadEvents = new SimpleBooleanProperty(false);
+		this.scheduledTask = scheduler.scheduleWithFixedDelay(() -> {
+			if (fsEventAggregator.hasUpdates()) {
+				flush();
+			}
+		}, 1000, 1000, TimeUnit.MILLISECONDS);
+	}
```
The task will be canceled when the executorService shuts down.

infeo commented on src/main/java/org/cryptomator/event/FileSystemEventAggregator.java:
```diff
@@ -0,0 +1,101 @@
+	 *
+	 * @param target collection which is first cleared and then the EntrySet copied to.
+	 */
+	public void cloneTo(Collection<Map.Entry<FSEventBucket, FSEventBucketContent>> target) {
+		hasUpdates.set(false);
+		target.clear();
+		target.addAll(map.entrySet());
+	}
```
Fixed in https://github.com/cryptomator/cryptomator/pull/3780/commits/a93ceb5c8c52bf819c3ef2a8bc7b4175a0c0ef81

infeo commented:
@overheadhunter Ready for review. I updated the PR description to reflect the changes and also added some short java doc strings to the "model" classes.

overheadhunter commented on src/main/java/org/cryptomator/ui/fxapp/FxFSEventList.java:
```diff
@@ -0,0 +1,70 @@
+		this.events = FXCollections.observableArrayList();
+		this.eventAggregator = fsEventAggregator;
+		this.unreadEvents = new SimpleBooleanProperty(false);
+		this.scheduledTask = scheduler.scheduleWithFixedDelay(() -> {
+			if (fsEventAggregator.hasMaybeUpdates()) {
+				flush();
+			}
+		}, 1000, 1000, TimeUnit.MILLISECONDS);
```
instead of "pushing" events from background thread (scheduler) to ui thread, rather "pull" events from within the UI thread.

Create a [ScheduledService](https://openjfx.io/javadoc/21/javafx.graphics/javafx/concurrent/ScheduledService.htm), which will run jobs on the UI thread, so you don't need any `Platform.runLater` and no CDL to wait for the job to finish.

infeo replied:
> Create a [ScheduledService](https://openjfx.io/javadoc/21/javafx.graphics/javafx/concurrent/ScheduledService.htm), which will run jobs on the UI thread

You mistake the purpose of a ScheduleService. It is not intended for running things on the UI thread. It is only intended to be observable. Easy test is:

```java
ScheduledService<Void> svc = new ScheduledService<Void>() {
    protected Task<Void> createTask() {
        return new Task<Void>() {
            protected Void call() {
            	if(!Platform.isFxApplicationThread()) {
		    System.out.println("NOT FX THREAD");
		}
                return null;
            }
        };
    }
};
 svc.setPeriod(Duration.seconds(1));
 svc.start();
```

This will print every second "NOT FX THREAD" to the console. See also the [Task](https://openjfx.io/javadoc/23/javafx.graphics/javafx/concurrent/Task.html#a-task-which-modifies-the-scene-graph-heading) documentation, there is also the Platform.runLater() method used.

infeo replied:
Of course, we can set the `executor` property. 😉

infeo replied:
Changed in https://github.com/cryptomator/cryptomator/pull/3780/commits/bb48551aacc814251276a292718a9da538cfcfcf

overheadhunter commented on src/main/java/org/cryptomator/ui/fxapp/FxApplication.java:
```diff
@@ -29,9 +29,10 @@ public class FxApplication {
 	private final FxApplicationStyle applicationStyle;
 	private final FxApplicationTerminator applicationTerminator;
 	private final AutoUnlocker autoUnlocker;
+	private final FxFSEventList fxFSEventList; //not unused! By injecting it here, the object gets initiated the service starts
```
me is not happy.

ok for now, but we might eventually want to create a meta background task class, which is the only thing injected in the fx main app.

infeo replied:
I changed it to a manual start. To reduce technical debt. (see https://github.com/cryptomator/cryptomator/pull/3780/commits/bb48551aacc814251276a292718a9da538cfcfcf)

overheadhunter reviewed (changes requested):
core impl is fine, but UI thread handling needs a little change

overheadhunter commented on src/main/java/org/cryptomator/ui/fxapp/FxFSEventList.java:
```diff
@@ -0,0 +1,77 @@
+
+		@Override
+		protected Void call() {
+			if (eventAggregator.hasMaybeUpdates()) {
+				eventAggregator.cloneTo(events);
+				unreadEvents.setValue(true);
+			}
```
we expect `hasMaybeUpdates()` to be false in most cases, right? In that case I'd rather run this task on a background thread and only call `Platform.runLater(...)` for the actual cloning/flagging.

infeo replied:
Done in https://github.com/cryptomator/cryptomator/pull/3780/commits/5b4245c257767fdc0964bf94cbdd98794e72055d

overheadhunter commented on src/main/java/org/cryptomator/ui/fxapp/FxFSEventList.java:
```diff
@@ -0,0 +1,68 @@
+		this.unreadEvents = new SimpleBooleanProperty(false);
+	}
+
+	public void startPolling() {
```
method name is a bit misleading, since it schedules just one single job

```suggestion
	public void scheduleUpdateCheck() {
```

overheadhunter reviewed (approved):
looks good from a technical point of view. Just a small naming suggestion
